kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt

```diff
@@ -213,3 +219,10 @@ internal inline fun DispatchedTask<*>.runUnconfinedEventLoop(
 internal inline fun Continuation<*>.resumeWithStackTrace(exception: Throwable) {
     resumeWith(Result.failure(recoverStackTrace(exception, this)))
 }
+
+/**
+ * This exception holds an exception raised in [CoroutineDispatcher.dispatch] method
+ *
+ * @see DispatchedContinuation.dispatchWithExceptionHandling
+ */
+internal class DispatchException(cause: Throwable) : Exception(cause)
```

Comment:
In theory, it is possible that the end user will see this exception. The simplest example:
```kotlin
val dispatcher = object: CoroutineDispatcher() {
    override fun dispatch(context: CoroutineContext, block: Runnable) {
        TODO("Not yet implemented")
    }

}
runBlocking {
    try {
        suspendCancellableCoroutine<Unit> { cont ->
            dispatcher.interceptContinuation(cont).resume(Unit)
        }
    } catch (e: Throwable) {
        println("Caught $e")
    }
}
```
This means that it's better to supply a diagnostic message, like, "a coroutine dispatcher threw an exception." It would also be nice to include the dispatcher (and possibly the rest of the coroutine context) in the message.